File: programmers/level3/네트워크.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> v[201];
bool visited[201];

void dfs(int cur){
	visited[cur] = true;
	for(unsigned int i=0; i<v[cur].size(); i++){
		if(!visited[v[cur][i]]){
			dfs(v[cur][i]);
		}
	}
}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    
    for(int i=0; i<n; i++){
		for(int j=0; j<n; j++){
			if(i < j && computers[i][j]){
				v[i].push_back(j);
				v[j].push_back(i);
			}
		}
	}
    
    for(int i=0; i<n; i++){
		if(!visited[i]){
			dfs(i);
			answer++;
		}
	}
    return answer;
}

File: programmers/level3/네트워크.js
function solution(n, computers) {
    let graph = Array.from(Array(n), () => []);
    let visited = [...Array(n).fill(false)];
    
    for(let i = 0; i < computers.length; i++){
        for(let j = i + 1; j < computers[i].length; j++){
            if(computers[i][j]){
                graph[i].push(j);
                graph[j].push(i);
            }
        }
    }
    
    const dfs = (cur) => {
        visited[cur] = true;
        for(let i = 0; i < graph[cur].length; i++){
            if(!visited[graph[cur][i]]){
                dfs(graph[cur][i]);
            }
        }
    }

    const bfs = (cur) => {
        const q = [];
        q.push(cur);
        visited[cur] = true;
        
        while(q.length !== 0){
            const x = q.shift();
            for(let i = 0; i < graph[x].length; i++){
                const next = graph[x][i];
                if(!visited[next]){
                    q.push(next);
                    visited[next] = true;
                }
            }
        }
    }
    
    let cnt = 0;
    for(let i = 0; i < n; i++){
        if(!visited[i]){
            dfs(i); //or bfs(i);
            cnt++;
        }
    }
    return cnt;
}
